contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,284 @@
+    address depositToken;
+    address tokenToBack;
+  }
+  address public protocol;
+  IPriceFeeds public priceFeed;
+  address public valuationToken;
+  address[] public stakingTokens;
+  mapping(address => bool) public tokenSupported;
+  mapping(uint256 => uint256) public sTokenPrice;
+  mapping(uint256 => TokenComposition) public identifierToTokens;
+  mapping(uint256 => uint256) public balanceStakedPerID;
+  mapping(uint256 => uint256) public supplyPerID;
+
+  mapping(address => uint256) public undistributedRewards;
+  mapping(uint256 => uint256) public rewardsPerToken;
+  address public rewardToken;
+  mapping(address => mapping(uint256 => uint256)) public lastClaimRewardAccrual;
+  mapping(uint256 => bool) public initialized;
+  mapping(address => uint8) public tokenDecimalCache;
+
+  event PriceFeed(address feed);
+  event Protocol(address protocol);
+  event ValuationToken(address token);
+  event TokensSupported(address[] supportedTokens);
+  event Deposit(address depositor, address tokenBacked, address depositToken, uint256 amount);
+  event Withdraw(address withdrawer, address tokenBacked, address receivedToken, uint256 amount);
+  event RewardToken(address token);
+  event PoolDraw(address tokenBacked, address tokenCovered, uint256 amountCovered);
+  event AddRewards(address tokenBacked, uint256 amount);
+  event AccumulateRewards(address tokenBacked, uint256 amount);
+  event DistributeRewards(address tokenBacked, uint256 amount);
+
+  modifier onlyProtocol() {
+    require(msg.sender == protocol, "not protocol");
+    _;
+  }
+
+  constructor(string memory uri_) ERC1155(uri_) {}
+
+  function updateTokenSupport(
+    address[] memory tokens,
+    bool[] memory support,
+    address[] memory listOfSupportedTokens
+  ) external onlyOwner {
+    stakingTokens = listOfSupportedTokens;
+    for (uint256 i; i < tokens.length; ) {
+      tokenSupported[tokens[i]] = support[i];
+      if (tokenDecimalCache[tokens[i]] == 0) {
+        tokenDecimalCache[tokens[i]] = IERC20Metadata(tokens[i]).decimals();
+      }
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function setRewardToken(address r) external onlyOwner {
+    rewardToken = r;
+    emit RewardToken(r);
+  }
+
+  function deposit(
+    address depositToken,
+    address tokenToBack,
+    uint256 amount
+  ) external {
+    require(tokenSupported[depositToken], "unsupported deposit token");
+    uint256 tokenID = convertToID(depositToken, tokenToBack);
+    if (identifierToTokens[tokenID].depositToken == address(0)) {
+      sTokenPrice[tokenID] = 1e18;
+    }
+    identifierToTokens[tokenID] = TokenComposition({depositToken: depositToken, tokenToBack: tokenToBack});
+    uint256 mintAmount = _amountToMint(tokenID, amount);
+    _claimReward(tokenID);
+    _mint(msg.sender, tokenID, mintAmount, "");
+    balanceStakedPerID[tokenID] += amount;
+    supplyPerID[tokenID] += mintAmount;
+    IERC20(depositToken).safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
worth moving safeTransfer couple of lines up? before the mint?

Reply:
This makes the code make more sense so will be done.

Reply:
implemented